[API-1543] [API-1651] [API-1622] [API-1701] Refactor

yuce opened the pull request:
TODO

harunalpak commented on internal/serialization/compact_field.go:
```diff
@@ -13,35 +13,205 @@ type CompactField struct {
+	serialization.FieldKindArrayOfFloat64: func(r serialization.CompactReader, field string) any {
+		return r.ReadArrayOfFloat64(field)
+	},
+	serialization.FieldKindString: func(r serialization.CompactReader, field string) any {
```
For FieldKindString we use "return readNullableCompactField(field, r.ReadString)". Is it correct?
( also for Decimal, Time, Date etc)

yuce replied:
Yes, those types are nullable.

harunalpak commented on internal/serialization/generic_compact.go:
```diff
@@ -77,55 +80,65 @@ func (cm GenericCompact) Read(reader serialization.CompactReader) interface{} {
+		serialization.FieldKindArrayOfDecimal:               reflect.TypeOf([]types.Decimal{}),
+		serialization.FieldKindTime:                         reflect.TypeOf(types.LocalTime(t)),
+		serialization.FieldKindArrayOfTime:                  reflect.TypeOf([]types.LocalTime{}),
+		serialization.FieldKindDate:                         reflect.TypeOf(types.LocalTime(t)),
```
I think there is small mistake. It could be "types.LocalDate(t)"?

yuce replied:
Thanks!

yuce replied:
Fixed at: https://github.com/hazelcast/hazelcast-commandline-client/pull/159/commits/ff8e67dd3673dbb32a0005c0509f3b81dc9944a4

harunalpak commented on internal/serialization/generic_compact.go:
```diff
@@ -77,55 +80,65 @@ func (cm GenericCompact) Read(reader serialization.CompactReader) interface{} {
+		serialization.FieldKindArrayOfTime:                  reflect.TypeOf([]types.LocalTime{}),
+		serialization.FieldKindDate:                         reflect.TypeOf(types.LocalTime(t)),
+		serialization.FieldKindArrayOfDate:                  reflect.TypeOf([]types.LocalDate{}),
+		serialization.FieldKindTimestamp:                    reflect.TypeOf(types.LocalTime(t)),
```
I think there is small mistake. It could be "types.LocalDateTime(t)"?

yuce replied:
Thanks!

yuce replied:
Fixed at: https://github.com/hazelcast/hazelcast-commandline-client/pull/159/commits/ff8e67dd3673dbb32a0005c0509f3b81dc9944a4

harunalpak commented on internal/serialization/generic_compact.go:
```diff
@@ -77,55 +80,65 @@ func (cm GenericCompact) Read(reader serialization.CompactReader) interface{} {
+		serialization.FieldKindArrayOfDate:                  reflect.TypeOf([]types.LocalDate{}),
+		serialization.FieldKindTimestamp:                    reflect.TypeOf(types.LocalTime(t)),
+		serialization.FieldKindArrayOfTimestamp:             reflect.TypeOf([]types.LocalDateTime{}),
+		serialization.FieldKindTimestampWithTimezone:        reflect.TypeOf(types.LocalTime(t)),
```
I think there is small mistake. It could be "types.OffsetDateTime(t)"?

yuce replied:
Thanks, fixed at: https://github.com/hazelcast/hazelcast-commandline-client/pull/159/commits/ff8e67dd3673dbb32a0005c0509f3b81dc9944a4

harunalpak commented on internal/serialization/generic_compact.go:
```diff
@@ -77,55 +80,65 @@ func (cm GenericCompact) Read(reader serialization.CompactReader) interface{} {
+		serialization.FieldKindArrayOfTimestampWithTimezone: reflect.TypeOf([]types.OffsetDateTime{}),
+		serialization.FieldKindCompact:                      reflect.TypeOf(a),
+		serialization.FieldKindArrayOfCompact:               reflect.TypeOf([]any{}),
+		serialization.FieldKindNullableBoolean:              reflect.TypeOf(&b),
```
I am a bit confusing about NullableBoolean => &b (address of boolean) ?

yuce replied:
Nullable boolean is a pointer, so we need a boolean and take its address to be able to get its type.

harunalpak commented on internal/serialization/serialization_consts.go:
```diff
@@ -100,155 +100,85 @@ const (
+	TypeByteArray:                            "BYTE_ARRAY",
+	TypeBoolArray:                            "BOOL_ARRAY",
+	TypeUInt16Array:                          "UINT16_ARRAY",
+	TypeInt16Array:                           "UINT16_ARRAY",
```
It could be "INT16_ARRAY"?

yuce replied:
Thanks, fixed at: https://github.com/hazelcast/hazelcast-commandline-client/pull/159/commits/ff8e67dd3673dbb32a0005c0509f3b81dc9944a4

harunalpak commented on internal/serialization/serialization_consts.go:
```diff
@@ -100,155 +100,85 @@ const (
+	TypeGobSerialization:                     "GO_GOB_SERIALIZATION",
+	TypeHibernate3TypeHibernateCacheKey:      "HIBERNATE3_CACHE_KEY",
+	TypeHibernate3TypeHibernateCacheEntry:    "HIBERNATE3_CACHE_ENTRY",
+	TypeHibernate4TypeHibernateCacheKey:      "HIBERNATE3_CACHE_KEY",
```
It could be "HIBERNATE4_CACHE_KEY"

yuce replied:
Thanks, fixed at: https://github.com/hazelcast/hazelcast-commandline-client/pull/159/commits/ff8e67dd3673dbb32a0005c0509f3b81dc9944a4

yuce commented on base/commands/config/config_list.go:
```diff
@@ -0,0 +1,59 @@
+
+func (cm ListCmd) Init(cc plug.InitContext) error {
+	cc.SetCommandUsage("list")
+	long := fmt.Sprintf(`Lists known configurations
```
The indentation is as-is, so if we indent those lines in the source, they would also appear indented in the output.

yuce commented on base/commands/sql/common.go:
```diff
@@ -0,0 +1,156 @@
+			// If Go client cannot find a connection, it returns immediately with ErrIO
+			// Retry logic here
+			if err != nil {
+				if errors.Is(err, hzerrors.ErrIO) {
```
Since this is a user tool, it makes more sense to me to retry indefinitely, until the user cancels manually.

yuce commented on base/commands/sql/common.go:
```diff
@@ -0,0 +1,156 @@
+			if _, err := ci.Client().SQL().Execute(ctx, serr.Suggestion); err != nil {
+				return nil, err
+			}
+			// execute the original query
```
Hazelcast can suggest a mapping for a query. CLC prints that suggestion, but if the user uses the `--use-mapping-suggestion` flag, then it automatically applies the suggestion and runs the errored query again. So, the user doesn't have to run the query again manually.

yuce commented on base/commands/sql/common.go:
```diff
@@ -0,0 +1,156 @@
+}
+
+func convertSQLType(ct sql.ColumnType) int32 {
+	t, ok := sqlTypeToSerializationType[ct]
```
`t` is the serialization type. `ok` is `true` if `ct` was found in `sqlTypeToSerializationType`. I can change `ok` to `found` if it makes the purpose more obvious.

yuce commented on base/commands/sql/sql.go:
```diff
@@ -0,0 +1,66 @@
+	cc.SetPositionalArgCount(1, 1)
+	cc.AddCommandGroup("sql", "SQL")
+	cc.SetCommandGroup("sql")
+	long := `Runs the given SQL query or starts the SQL shell
```
No, that would also indent the output.

yuce commented on clc/cmd/exec_context.go:
```diff
@@ -103,7 +104,7 @@ func (ec *ExecContext) ClientInternal(ctx context.Context) (*hazelcast.ClientInt
 }
 
 func (ec *ExecContext) Interactive() bool {
```
`Get` and `Is` are usually not prefixed to getters according to he Go naming conventions.

yuce commented on clc/cmd/imports.go:
```diff
@@ -0,0 +1,5 @@
+package cmd
```
The `init` functions in the packages which register plugins wouldn't run otherwise.

yuce commented on base/commands/config/config_import.go:
```diff
@@ -0,0 +1,236 @@
+	if ok {
+		return nil
+	}
+	ok, err = cm.tryImportViridianZipSource(ctx, ec, src)
```
This function tries to import the given string using two different strategies. I've added some comments at: https://github.com/hazelcast/hazelcast-commandline-client/pull/159/commits/ff8e67dd3673dbb32a0005c0509f3b81dc9944a4 Would that be adequate?

yuce commented on base/commands/dummy.go:
```diff
@@ -0,0 +1 @@
+package commands
```
Added at: https://github.com/hazelcast/hazelcast-commandline-client/pull/159/commits/ff8e67dd3673dbb32a0005c0509f3b81dc9944a4

yuce commented on base/commands/config/config_list.go:
```diff
@@ -0,0 +1,59 @@
+				Value: e.Name(),
+			}})
+		}
+	}
```
I've changed that code so a message is printed in the interactive mode at: https://github.com/hazelcast/hazelcast-commandline-client/pull/159/commits/ff8e67dd3673dbb32a0005c0509f3b81dc9944a4 The non-interactive mode doesn't show any error, but the error is logged.

harunalpak commented:
First of all, I tried to read the code as much as I could. But of course, since I only have basic knowledge of go, I won't comment on too much code. So I will do more functional testing. And you can see my comments below, which I found as a result of these functional tests. It has been a really big change and in general "clc" is very stable and works well. Nice job, well done.

harunalpak commented:
If our configured server was closed, and when I try to send a command to server, clc waiting for a long time. May be we can try a limited time and give an error that "can not reach the server". It can be more useful I think.

harunalpak commented:
For interactive mode "\" char can be a bit compelling. I think SQL commands are limited, and may be clc automatically detect these commands.

harunalpak commented:
For interactive mode when I wrore \help, I got the following message;

```
Usage:
   [command]

Distributed Data Structures
  map         Map operations
  objects     Generic distributed data structure operations

SQL
  sql         Run SQL

Additional Commands:
  completion  Generate the autocompletion script for the specified shell
  config      Show, add or change configuration
  help        Help about any command
  home        Print the CLC home directory, optionally by joining the given sub-path
  version     Print CLC version

Flags:
  -f, --format string   set the output type, one of: csv, delimited, json, table (default "table")
  -h, --help            help for this command
      --verbose         enable verbose output
```
But this is non-interactive mode help text. So it is a bit confusing. Can we change it?

harunalpak commented:
In interactive mode, when I wrote "help", the following text shown;

```
Error: 
Shortcut Commands:
        \dm           List mappings
        \dm  MAPPING  Display information about a mapping
        \dm+ MAPPING  Describe a mapping
        \exit         Exit the shell
        \help         Display help for CLC commands
```

This is good. But "Error:" could be remove.

harunalpak commented:
For interactive mode help, could we add "\home"?

```
Shortcut Commands:
        \dm           List mappings
        \dm  MAPPING  Display information about a mapping
        \dm+ MAPPING  Describe a mapping
        \exit         Exit the shell
        \help         Display help for CLC commands
```

harunalpak commented:
Can we remove "completion"  command from help?

I tried on interactive mode "\completion bash" and I got the following message;

```
# bash completion V2 for                                      -*- shell-script -*-

___debug()
{
    if [[ -n ${BASH_COMP_DEBUG_FILE:-} ]]; then
        echo "$*" >> "${BASH_COMP_DEBUG_FILE}"
    fi
}

# Macs have bash3 for which the bash-completion package doesn't include
# _init_completion. This is a minimal version of that function.
___init_completion()
{
    COMPREPLY=()
    _get_comp_words_by_ref "$@" cur prev words cword
}

# This function calls the  program to obtain the completion
# results and the directive.  It fills the 'out' and 'directive' vars.
___get_completion_results() {
    local requestComp lastParam lastChar args

    # Prepare the command to request completions for the program.
    # Calling ${words[0]} instead of directly  allows to handle aliases
    args=("${words[@]:1}")
    requestComp="${words[0]} __complete ${args[*]}"

    lastParam=${words[$((${#words[@]}-1))]}
    lastChar=${lastParam:$((${#lastParam}-1)):1}
    ___debug "lastParam ${lastParam}, lastChar ${lastChar}"

    if [ -z "${cur}" ] && [ "${lastChar}" != "=" ]; then
        # If the last parameter is complete (there is a space following it)
        # We add an extra empty parameter so we can indicate this to the go method.
        ___debug "Adding extra empty parameter"
        requestComp="${requestComp} ''"
    fi

    # When completing a flag with an = (e.g.,  -n=<TAB>)
    # bash focuses on the part after the =, so we need to remove
    # the flag part from $cur
    if [[ "${cur}" == -*=* ]]; then
        cur="${cur#*=}"
    fi

    ___debug "Calling ${requestComp}"
    # Use eval to handle any environment variables and such
    out=$(eval "${requestComp}" 2>/dev/null)

    # Extract the directive integer at the very end of the output following a colon (:)
    directive=${out##*:}
    # Remove the directive
    out=${out%:*}
    if [ "${directive}" = "${out}" ]; then
        # There is not directive specified
        directive=0
    fi
    ___debug "The completion directive is: ${directive}"
    ___debug "The completions are: ${out}"
}

___process_completion_results() {
    local shellCompDirectiveError=1
    local shellCompDirectiveNoSpace=2
    local shellCompDirectiveNoFileComp=4
    local shellCompDirectiveFilterFileExt=8
    local shellCompDirectiveFilterDirs=16

    if [ $((directive & shellCompDirectiveError)) -ne 0 ]; then
        # Error code.  No completion.
        ___debug "Received error from custom completion go code"
        return
    else
        if [ $((directive & shellCompDirectiveNoSpace)) -ne 0 ]; then
            if [[ $(type -t compopt) = "builtin" ]]; then
                ___debug "Activating no space"
                compopt -o nospace
            else
                ___debug "No space directive not supported in this version of bash"
            fi
        fi
        if [ $((directive & shellCompDirectiveNoFileComp)) -ne 0 ]; then
            if [[ $(type -t compopt) = "builtin" ]]; then
                ___debug "Activating no file completion"
                compopt +o default
            else
                ___debug "No file completion directive not supported in this version of bash"
            fi
        fi
    fi

    # Separate activeHelp from normal completions
    local completions=()
    local activeHelp=()
    ___extract_activeHelp

    if [ $((directive & shellCompDirectiveFilterFileExt)) -ne 0 ]; then
        # File extension filtering
        local fullFilter filter filteringCmd

        # Do not use quotes around the $completions variable or else newline
        # characters will be kept.
        for filter in ${completions[*]}; do
            fullFilter+="$filter|"
        done

        filteringCmd="_filedir $fullFilter"
        ___debug "File filtering command: $filteringCmd"
        $filteringCmd
    elif [ $((directive & shellCompDirectiveFilterDirs)) -ne 0 ]; then
        # File completion for directories only

        # Use printf to strip any trailing newline
        local subdir
        subdir=$(printf "%s" "${completions[0]}")
        if [ -n "$subdir" ]; then
            ___debug "Listing directories in $subdir"
            pushd "$subdir" >/dev/null 2>&1 && _filedir -d && popd >/dev/null 2>&1 || return
        else
            ___debug "Listing directories in ."
            _filedir -d
        fi
    else
        ___handle_completion_types
    fi

    ___handle_special_char "$cur" :
    ___handle_special_char "$cur" =

    # Print the activeHelp statements before we finish
    if [ ${#activeHelp[*]} -ne 0 ]; then
        printf "\n";
        printf "%s\n" "${activeHelp[@]}"
        printf "\n"

        # The prompt format is only available from bash 4.4.
        # We test if it is available before using it.
        if (x=${PS1@P}) 2> /dev/null; then
            printf "%s" "${PS1@P}${COMP_LINE[@]}"
        else
            # Can't print the prompt.  Just print the
            # text the user had typed, it is workable enough.
            printf "%s" "${COMP_LINE[@]}"
        fi
    fi
}

# Separate activeHelp lines from real completions.
# Fills the $activeHelp and $completions arrays.
___extract_activeHelp() {
    local activeHelpMarker="_activeHelp_ "
    local endIndex=${#activeHelpMarker}

    while IFS='' read -r comp; do
        if [ "${comp:0:endIndex}" = "$activeHelpMarker" ]; then
            comp=${comp:endIndex}
            ___debug "ActiveHelp found: $comp"
            if [ -n "$comp" ]; then
                activeHelp+=("$comp")
            fi
        else
            # Not an activeHelp line but a normal completion
            completions+=("$comp")
        fi
    done < <(printf "%s\n" "${out}")
}

___handle_completion_types() {
    ___debug "___handle_completion_types: COMP_TYPE is $COMP_TYPE"

    case $COMP_TYPE in
    37|42)
        # Type: menu-complete/menu-complete-backward and insert-completions
        # If the user requested inserting one completion at a time, or all
        # completions at once on the command-line we must remove the descriptions.
        # https://github.com/spf13/cobra/issues/1508
        local tab=$'\t' comp
        while IFS='' read -r comp; do
            [[ -z $comp ]] && continue
            # Strip any description
            comp=${comp%%$tab*}
            # Only consider the completions that match
            if [[ $comp == "$cur"* ]]; then
                COMPREPLY+=("$comp")
            fi
        done < <(printf "%s\n" "${completions[@]}")
        ;;

    *)
        # Type: complete (normal completion)
        ___handle_standard_completion_case
        ;;
    esac
}

___handle_standard_completion_case() {
    local tab=$'\t' comp

    # Short circuit to optimize if we don't have descriptions
    if [[ "${completions[*]}" != *$tab* ]]; then
        IFS=$'\n' read -ra COMPREPLY -d '' < <(compgen -W "${completions[*]}" -- "$cur")
        return 0
    fi

    local longest=0
    local compline
    # Look for the longest completion so that we can format things nicely
    while IFS='' read -r compline; do
        [[ -z $compline ]] && continue
        # Strip any description before checking the length
        comp=${compline%%$tab*}
        # Only consider the completions that match
        [[ $comp == "$cur"* ]] || continue
        COMPREPLY+=("$compline")
        if ((${#comp}>longest)); then
            longest=${#comp}
        fi
    done < <(printf "%s\n" "${completions[@]}")

    # If there is a single completion left, remove the description text
    if [ ${#COMPREPLY[*]} -eq 1 ]; then
        ___debug "COMPREPLY[0]: ${COMPREPLY[0]}"
        comp="${COMPREPLY[0]%%$tab*}"
        ___debug "Removed description from single completion, which is now: ${comp}"
        COMPREPLY[0]=$comp
    else # Format the descriptions
        ___format_comp_descriptions $longest
    fi
}

___handle_special_char()
{
    local comp="$1"
    local char=$2
    if [[ "$comp" == *${char}* && "$COMP_WORDBREAKS" == *${char}* ]]; then
        local word=${comp%"${comp##*${char}}"}
        local idx=${#COMPREPLY[*]}
        while [[ $((--idx)) -ge 0 ]]; do
            COMPREPLY[$idx]=${COMPREPLY[$idx]#"$word"}
        done
    fi
}

___format_comp_descriptions()
{
    local tab=$'\t'
    local comp desc maxdesclength
    local longest=$1

    local i ci
    for ci in ${!COMPREPLY[*]}; do
        comp=${COMPREPLY[ci]}
        # Properly format the description string which follows a tab character if there is one
        if [[ "$comp" == *$tab* ]]; then
            ___debug "Original comp: $comp"
            desc=${comp#*$tab}
            comp=${comp%%$tab*}

            # $COLUMNS stores the current shell width.
            # Remove an extra 4 because we add 2 spaces and 2 parentheses.
            maxdesclength=$(( COLUMNS - longest - 4 ))

            # Make sure we can fit a description of at least 8 characters
            # if we are to align the descriptions.
            if [[ $maxdesclength -gt 8 ]]; then
                # Add the proper number of spaces to align the descriptions
                for ((i = ${#comp} ; i < longest ; i++)); do
                    comp+=" "
                done
            else
                # Don't pad the descriptions so we can fit more text after the completion
                maxdesclength=$(( COLUMNS - ${#comp} - 4 ))
            fi

            # If there is enough space for any description text,
            # truncate the descriptions that are too long for the shell width
            if [ $maxdesclength -gt 0 ]; then
                if [ ${#desc} -gt $maxdesclength ]; then
                    desc=${desc:0:$(( maxdesclength - 1 ))}
                    desc+="…"
                fi
                comp+="  ($desc)"
            fi
            COMPREPLY[ci]=$comp
            ___debug "Final comp: $comp"
        fi
    done
}

__start_()
{
    local cur prev words cword split

    COMPREPLY=()

    # Call _init_completion from the bash-completion package
    # to prepare the arguments properly
    if declare -F _init_completion >/dev/null 2>&1; then
        _init_completion -n "=:" || return
    else
        ___init_completion -n "=:" || return
    fi

    ___debug
    ___debug "========= starting completion logic =========="
    ___debug "cur is ${cur}, words[*] is ${words[*]}, #words[@] is ${#words[@]}, cword is $cword"

    # The user could have moved the cursor backwards on the command-line.
    # We need to trigger completion from the $cword location, so we need
    # to truncate the command-line ($words) up to the $cword location.
    words=("${words[@]:0:$cword+1}")
    ___debug "Truncated words[*]: ${words[*]},"

    local out directive
    ___get_completion_results
    ___process_completion_results
}

if [[ $(type -t compopt) = "builtin" ]]; then
    complete -o default -F __start_ 
else
    complete -o default -o nospace -F __start_ 
fi

# ex: ts=4 sw=4 et filetype=sh
```

yuce commented:
Thanks for the review!

> If our configured server was closed, and when I try to send a command to server, clc waiting for a long time. May be we can try a limited time and give an error that "can not reach the server". It can be more useful I think.

In the non-interactive mode, we don't display any message but the log messages are written to files in `$CLC_HOME/logs`. The user also has the option to change the log file using `--log.path` and can use `--log.path stderr` to enable logging to stderr.

In the interactive mode we display a message that the CLC is connecting to the cluster. The user has the option to change log file like the non-interactivemode.

I'll update our documentation with that information.

> For interactive mode "" char can be a bit compelling. I think SQL commands are limited, and may be clc automatically detect these commands.
For now we'll go with required `\` but let's re-consider removing that in the future.

> For interactive mode when I wrore \help, I got the following message;
> But this is non-interactive mode help text. So it is a bit confusing. Can we change it?

Thanks, I've updated the help message in the interactive mode to have the `\` prefix at: https://github.com/hazelcast/hazelcast-commandline-client/pull/159/commits/6bf1c96856f57a9da9acb3a208afd9e1a274b2e0

> In interactive mode, when I wrote "help", the following text shown;
> This is good. But "Error:" could be remove.

Fixed at: https://github.com/hazelcast/hazelcast-commandline-client/pull/159/commits/6bf1c96856f57a9da9acb3a208afd9e1a274b2e0

> Also we use MAPPING in descriptions. But we import a SQL table. So it can be confusing.
We don't have tables on the Hazelcast side, just mappings.

> In addition when I check \dm+ cities
> In this table I was used table_catalog, table_schema... So we can use table instead of mapping.
The column names are coming directly from the server. `\dm+`  just runs:
```
SELECT * FROM information_schema.columns
WHERE table_name = '%s'
```

> For interactive mode help, could we add "\home" and "\version"?
Those are available in the `\help` command.

> Can we remove "completion" command from help?
That makes sense. I've removed it from the interactive mode at: https://github.com/hazelcast/hazelcast-commandline-client/pull/159/commits/6bf1c96856f57a9da9acb3a208afd9e1a274b2e0

yuce commented:
> Can we remove \ from command?

That's coming from the flag library we are using. It's probably trying to escape the backslash, it's not trivial to change that.

harunalpak commented:
For non-interactive mode help contains "completion  Generate the autocompletion script for the specified shell" also.

Can we remove "completion" from help?

And also when I wrote a command "completion bash" gives me the following;

```
harunalpak@Haruns-MacBook-Pro build % ./clc completion bash
# bash completion V2 for clc                                  -*- shell-script -*-

__clc_debug()
{
    if [[ -n ${BASH_COMP_DEBUG_FILE:-} ]]; then
        echo "$*" >> "${BASH_COMP_DEBUG_FILE}"
    fi
}

# Macs have bash3 for which the bash-completion package doesn't include
# _init_completion. This is a minimal version of that function.
__clc_init_completion()
{
    COMPREPLY=()
    _get_comp_words_by_ref "$@" cur prev words cword
}
```

Can we fix it?

yuce commented:
> For non-interactive mode help contains "completion Generate the autocompletion script for the specified shell" also.
> Can we remove "completion" from help?

No, that's how the user generates autocompletion for their shell (bash, powershell, etc.).

> And also when I wrote a command "completion bash" gives me the following;

The user is supposed to redirect the output to a file, so that's the correct behaviour:
```
$ clc completion bash > clc-autocomplete.bash
# or
$ source <(clc completion bash)
```
`clc completion bash --help` contains more information
